src/Parser.php

```diff
@@ -111,11 +111,18 @@ public function parse(string &$data, string $encoding, string $url = '')
             $declaration = $this->registry->create(DeclarationParser::class, [substr($data, 5, $pos - 5)]);
             if ($declaration->parse()) {
                 $data = substr($data, $pos + 2);
-                $data = '<?xml version="' . $declaration->version . '" encoding="' . $encoding . '" standalone="' . (($declaration->standalone) ? 'yes' : 'no') . '"?>' ."\n". $this->declare_html_entities() . $data;
+                $data = preg_replace('/^\\s*<!DOCTYPE\\s[^>\\[\\]]*>\s*/', '', $data);  // Strip DOCTYPE except if containing an [internal subset]
+                $data = '<?xml version="' . $declaration->version . '" encoding="' . $encoding . '" standalone="' . (($declaration->standalone) ? 'yes' : 'no') . '"?>' . "\n" .
+                    (preg_match('/^\\s*<!DOCTYPE\\s/', $data) ? '' : $this->declare_html_entities()) .  // Declare HTML entities only if no remaining DOCTYPE
+                    $data;
             } else {
                 $this->error_string = 'SimplePie bug! Please report this!';
                 return false;
             }
+        } else {
+            $data = preg_replace('/^\\s*<!DOCTYPE\\s[^>\\[\\]]*>\s*/', '', $data);   // Strip DOCTYPE except if containing an [internal subset]
+            $data = (preg_match('/^\\s*<!DOCTYPE\\s/', $data) ? '' : $this->declare_html_entities()) .  // Declare HTML entities only if no remaining DOCTYPE
```

Comment:
Instead of second matching, we could check the `$count` out-argument of `preg_replace`. But it probably does not matter in practice.

Reply:
I am not sure how to do that, since we may have three different cases:
* No DTD at all
* A DTD that should be removed
* A DTD that should not be removed